src/include/buffer.h

```diff
@@ -411,22 +412,32 @@ namespace buffer CEPH_BUFFER_API {
     void push_front(ptr& bp) {
       if (bp.length() == 0)
 	return;
+      _len += bp.length();
       _buffers.push_front(bp);
+    }
+    void push_front(ptr&& bp) {
+      if (bp.length() == 0)
+	return;
+      _buffers.push_front(std::move(bp));
       _len += bp.length();
     }
     void push_front(raw *r) {
-      ptr bp(r);
-      push_front(bp);
+      push_front(ptr(r));
     }
     void push_back(const ptr& bp) {
       if (bp.length() == 0)
 	return;
+      _len += bp.length();
       _buffers.push_back(bp);
+    }
+    void push_back(ptr&& bp) {
+      if (bp.length() == 0)
+	return;
+      _buffers.push_back(std::move(bp));
```

Comment:
same here. you moved the call to bp.length() in `void push_back(const ptr& bp) {` instead